executor/src/subscriber.rs

```diff
@@ -120,144 +153,155 @@ impl Subscriber {
-                        Error::permanent(PayloadRetrieveError(certificate_id, err.to_string()))
-                    })?;
+    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
+    async fn try_fetch_locally(&self, digest: BatchDigest, worker_id: WorkerId) -> Option<Batch> {
+        let _timer = self.metrics.subscriber_local_fetch_latency.start_timer();
+        let worker = self
+            .worker_cache
+            .load()
+            .worker(&self.name, &worker_id)
+            .expect("Own worker not found in cache")
+            .name;
+        let payload = self.network.get_payload(&worker, digest).await;
+        match payload {
+            Ok(Some(batch)) => {
+                debug!("Payload {} found locally", digest);
+                self.metrics.subscriber_local_hit.inc();
+                return Some(batch);
+            }
+            Ok(None) => debug!("Payload {} not found locally", digest),
+            Err(err) => error!("Error communicating with out own worker: {}", err),
+        }
+        None
+    }
 
-                match receiver.await.map_err(|err| {
-                    Error::permanent(PayloadRetrieveError(certificate_id, err.to_string()))
-                })? {
-                    Ok(block) => {
-                        // we successfully received the payload. Now let's add to store
-                        batch_store
-                            .write_all(
-                                block
-                                    .batches
-                                    .into_iter()
-                                    .map(|b| ((certificate_id, b.id), b.transactions)),
-                            )
-                            .await
-                            .map_err(|err| Error::permanent(SubscriberError::from(err)))?;
+    /// This future performs fetch from given worker
+    /// This future performs infinite retries with exponential backoff
+    /// You can specify stagger_delay before request is issued
+    #[instrument(level = "debug", skip_all, fields(stagger_delay = ? stagger_delay, worker = % worker, digest = % digest, authority = % authority))]
+    async fn fetch_from_worker(
+        &self,
+        stagger_delay: Duration,
+        worker: NetworkPublicKey,
+        digest: BatchDigest,
+        authority: PublicKey,
+    ) -> Batch {
+        tokio::time::sleep(stagger_delay).await;
+        let max_timeout = Duration::from_secs(60);
+        let mut timeout = Duration::from_secs(10);
+        loop {
+            let deadline = Instant::now() + timeout;
+            let get_payload_guard =
+                PendingGuard::make_inc(&self.metrics.pending_remote_get_payload);
+            let payload =
+                tokio::time::timeout_at(deadline, self.network.get_payload(&worker, digest)).await;
+            drop(get_payload_guard);
+            match payload {
+                Ok(Ok(Some(payload))) => return payload,
```

Comment:
Should we verify the payload actually corresponds to the digest here? I'm not sure if this verification is done at a higher or lower level.